src/main.rs

```diff
@@ -3,25 +3,42 @@ mod stake_accounts;
 
 use crate::args::{parse_args, Command};
 use crate::stake_accounts::{move_stake_account, TransferStakeKeys};
+use solana_cli_config::Config;
 use solana_client::thin_client::create_client;
 use solana_core::{cluster_info::VALIDATOR_PORT_RANGE, gossip_service::discover_cluster};
 use solana_sdk::{pubkey::Pubkey, signature::Keypair};
 use std::env;
+use std::net::SocketAddr;
+use url::Url;
 
-fn main() {
-    let config = parse_args(env::args_os());
-    let (nodes, _) = discover_cluster(&config.entrypoint_addr, config.num_nodes).unwrap();
-    let mut target = None;
-    for node in &nodes {
-        if node.gossip == config.entrypoint_addr {
-            target = Some(node.client_facing_addr());
-            break;
+fn compute_entrypoint_addr(json_rpc_url: &str) -> Vec<SocketAddr> {
+    let mut url = json_rpc_url.parse::<Url>().unwrap();
+    url.set_port(Some(8001)).unwrap();
+    url.socket_addrs(|| None).unwrap()
+}
+
+// TODO: This is unreliable when pointed at http://api.mainnet-beta.solana.com
```

Comment:
@sakridge, @mvines, I don't know the preferred way to connect to the cluster.  This method works well when targeting http://devnet.solana.com, but fails to find the gossip_addr about 50% of the time when targeting http://api.mainnet-beta.solana.com.  Am I doing something obviously wrong here?

Reply:
What was num_nodes in that case? If you passed a low value, then it may have returned early before it got the entry for the one you are targetting.

Reply:
Why not use RPC?

Reply:
@sakridge, num_nodes was 1, but I also set it to 4, and it was still unreliable.

@mvines, I'm looking for an implementation of the `Client` trait. To my knowledge, only `ThinClient` implements it.  Also, and maybe it doesn't matter, but I'm interested in maximizing performance. Ideally, I'd send 25 transactions all at once and then check their statuses afterward. Can I do that with RPC or is it more of a one-at-time interface?

Reply:
Give https://github.com/solana-labs/solana/blob/2623c71ed35e86bc7aeb4cbf74a12bf1f8eabb21/client/src/rpc_client.rs#L391 a try. 

Messing around in gossip for an end-user program is just going to hurt.  All firewalls/NATs/switches know how to HTTP across them.    `ThinClient` is an implementation detail of bench-tps as far as I'm concerned.

Reply:
Thanks @mvines, nice and reliable now.  That dropped a lot of build dependencies too.  I'll go with RpcClient for now, but Seems weird to not send transactions to a Transaction Processing Unit. And not using `Client` feels awful (means I can't unit-test with BankClient), but maybe an excuse to see if RpcClient could be implementing it.

Reply:
> num_nodes was 1, but I also set it to 4, and it was still unreliable.

Just for future reference, discover_cluster works by connecting to gossip and looking for num_nodes. Once it reaches that number, it will return. So in this case, if you are looking for a specific node, you would have to pass the value of all the nodes in the network to make sure you get the node that you are looking for and not some subset that didn't contain it. This interface is probably not the best for this use-case.

If you know (or the user of the CLI knows) the rpc port and node they want to talk to, then yes just directly create an RpcClient or ThinClient to that node directly, this is unrelated to the gossip part though.

I don't get the implementation detail of bench-tps comment. I see that RpcClient is one backend for ThinClient, BankClient being another. ThinClient gives you somethings that many clients might want which is finding the best node to talk to and some more robust retry interfaces for transactions. Seems also having the ability to unit-test with BankClient is reason enough to use ThinClient, but dunno..

Reply:
@sakridge, Thanks for this detailed reply. Super helpful!